File: memory-domain/src/main/java/com/memory/dto/comment/CommentResponse.java
package com.memory.dto.comment;

import com.memory.domain.comment.Comment;
import com.memory.dto.member.response.MemberResponse;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class CommentResponse {

    private Long id;
    private String content;
    private Integer depth;
    private Long memoryId;
    private MemberResponse member;
    private Long parentCommentId;
    private List<CommentResponse> children;
    private Long childrenCount;
    private LocalDateTime createDate;
    private LocalDateTime updateDate;

    private Boolean isDeleted;
    private Boolean isAuthor;

    public static CommentResponse from(Comment comment) {
        return from(comment, null, false);
    }

    public static CommentResponse from(Comment comment, Long currentMemberId, boolean includeChildren) {
        MemberResponse memberResponse = MemberResponse.from(comment.getMember());
        Long parentCommentId = comment.getParent() != null ? comment.getParent().getId() : null;

        List<CommentResponse> childrenResponses = includeChildren
                ? comment.getChildren().stream()
                        .map(child -> CommentResponse.from(child, currentMemberId, true))
                        .collect(Collectors.toList())
                : null;

        return new CommentResponse(
                comment.getId(),
                comment.getContent(),
                comment.getDepth(),
                comment.getMemory().getId(),
                memberResponse,
                parentCommentId,
                childrenResponses,
                (long) (childrenResponses != null ? childrenResponses.size() : 0),
                comment.getCreateDate(),
                comment.getUpdateDate(),
                comment.isDeleted(),
                comment.getMember().getId().equals(currentMemberId)
        );
    }

    public static List<CommentResponse> fromList(List<Comment> comments, Long currentMemberId) {
        return comments.stream()
                .map(comment -> CommentResponse.from(comment, currentMemberId, true))
                .collect(Collectors.toList());
    }
}

File: memory-api/src/main/kotlin/com/memory/dto/comment/CommentResponse.kt
package com.memory.dto.comment

import com.memory.domain.comment.Comment
import com.memory.dto.member.response.MemberResponse
import java.time.LocalDateTime

data class CommentResponse private constructor(
    val id: Long?,
    val content: String,
    val depth: Int,
    val memoryId: Long?,
    val member: MemberResponse,
    val parentCommentId: Long?,
    val children: List<CommentResponse>?,
    val childrenCount: Long,
    val createDate: LocalDateTime?,
    val updateDate: LocalDateTime?,
    val isDeleted: Boolean,
    val isAuthor: Boolean
) {
    companion object {
        @JvmStatic
        fun from(comment: Comment): CommentResponse =
            from(comment, null, false)

        @JvmStatic
        fun from(
            comment: Comment,
            currentMemberId: Long?,
            includeChildren: Boolean
        ): CommentResponse {
            val memberResponse = MemberResponse.from(comment.member)
            val parentId = comment.parent?.id

            val childrenResponses: List<CommentResponse>? =
                if (includeChildren)
                    comment.children.map { child -> from(child, currentMemberId, true) }
                else null

            val count = childrenResponses?.size?.toLong() ?: 0L
            val author = currentMemberId != null && comment.member.id == currentMemberId

            return CommentResponse(
                id = comment.id,
                content = comment.content,
                depth = comment.depth,
                memoryId = comment.memory.id,
                member = memberResponse,
                parentCommentId = parentId,
                children = childrenResponses,
                childrenCount = count,
                createDate = comment.createDate,
                updateDate = comment.updateDate,
                isDeleted = comment.isDeleted(),
                isAuthor = author
            )
        }

        @JvmStatic
        fun fromList(comments: List<Comment>, currentMemberId: Long?): List<CommentResponse> =
            comments.map { from(it, currentMemberId, true) }
    }
}
